File: 2023/src/three/day_three.py
import os

def is_adjacent(line, start_idx, number_length):
    min_bound = start_idx - 1
    max_bound = start_idx + number_length
    for idx, c in enumerate(line):
        if c != '.' and not c.isdigit():
            if idx >= min_bound and idx <= max_bound:
                return True
    return False

def is_symbol_around(lines, line_cnt, line_nr, start_idx, number_str):
    number_length = len(number_str)

    if line_nr == 0:
        current_line = lines[line_nr]
        line_below = lines[line_nr + 1]
        return is_adjacent(current_line, start_idx, number_length) or is_adjacent(line_below, start_idx, number_length)
    elif line_nr == line_cnt - 1:
        line_above = lines[line_nr - 1]
        current_line = lines[line_nr]
        return is_adjacent(line_above, start_idx, number_length) or is_adjacent(current_line, start_idx, number_length)
    elif line_nr > 0 and line_nr < line_cnt:
        line_above = lines[line_nr - 1]
        current_line = lines[line_nr]
        line_below = lines[line_nr + 1]
        return is_adjacent(line_below, start_idx, number_length) or is_adjacent(current_line, start_idx, number_length) or is_adjacent(line_above, start_idx, number_length)
    else:
        print("Invalid linenr {}".format(line_nr))
        return False

def find_part_numbers_w_idx(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()

    line_count = len(lines)

    result = []
    current_number_str = ""

    current_line_nr = 0

    last_idx = 0

    line_cntr = 0

    result_sum = 0

    for line in lines:
        line_cntr += 1
        result_line = []
        for idx, c in enumerate(line):
            if c.isdigit():
                current_number_str += c
            else:
                if current_number_str:
                    current_number = int(current_number_str)

                    start_num_idx = idx - len(current_number_str)

                    if is_symbol_around(lines, line_count, current_line_nr, start_num_idx, current_number_str):
                        result.append((current_number, start_num_idx))
                        result_line.append((current_number, start_num_idx))
                        result_sum += current_number
                    current_number_str = ""
            last_idx = idx

        print("Line {} {}\n".format(line_cntr, result_line))
        current_line_nr += 1

    print("{}".format(result_sum))
    return result

def day_three_part_one(filename):
    output = find_part_numbers_w_idx(filename)
    answer = sum([value for value, _ in output])
    return answer

filename = "D:/git/magic/aoc2023/aoc2023/src/three/input.txt"
day_three_part_one(filename)

# in rust: 560570, fout
# in python: 561307, fout
# met dezelfde code......

def is_special_char(char):
    #print("checking: ", char)
    return ((char != '.') and (not char.isdigit()))

# ff iq0
def has_sticky_symbol(lines, current_line_nr, char_idx):

    if current_line_nr == 0:
        current_line = lines[current_line_nr]
        line_below = lines[current_line_nr + 1]

        #print(current_line)
        #print(line_below)

        if char_idx == 0:
            #kijk rechts, rechts beneden, onder
            # 777..
            # .....
            # .....
            if (is_special_char(current_line[char_idx + 1])
            or is_special_char(line_below[char_idx])
            or is_special_char(line_below[char_idx + 1])):
                return True

        elif char_idx == (len(current_line) - 1):
            if (is_special_char(current_line[char_idx - 1])
            or is_special_char(line_below[char_idx] )
            or is_special_char(line_below[char_idx - 1] )):
                return True

        elif char_idx > 0 and char_idx < (len(current_line) - 1):
            if (is_special_char(current_line[char_idx - 1] )
            or is_special_char(current_line[char_idx + 1])
            or is_special_char(line_below[char_idx - 1] )
            or is_special_char(line_below[char_idx] )
            or is_special_char(line_below[char_idx + 1] )):
                return True
        
        # anders niks
        return False

    elif current_line_nr == (len(lines) - 1):
        line_above = lines[current_line_nr - 1]
        current_line = lines[current_line_nr]
        
        if char_idx == 0:
            # .....
            # .....
            # 777..
            if (is_special_char(current_line[char_idx + 1])
            or is_special_char(line_above[char_idx] )
            or is_special_char(line_above[char_idx + 1] )):
                return True

        elif char_idx == (len(current_line) - 1): # eind
            if (is_special_char(current_line[char_idx - 1])
            or is_special_char(line_above[char_idx] )
            or is_special_char(line_above[char_idx - 1] )):
                return True

        elif char_idx > 0 and char_idx < (len(current_line) - 1):
            if (is_special_char(current_line[char_idx - 1])
            or is_special_char(current_line[char_idx + 1] )
            or is_special_char(line_above[char_idx - 1] )
            or is_special_char(line_above[char_idx] )
            or is_special_char(line_above[char_idx + 1])):
                return True

        return False
            
    elif current_line_nr > 0 and current_line_nr < (len(lines) - 1):
        line_above = lines[current_line_nr - 1]
        current_line = lines[current_line_nr]
        line_below = lines[current_line_nr + 1]

        if char_idx == 0: #begin
            # .....
            # .....
            # 777..
            if (is_special_char(current_line[char_idx + 1])
            or is_special_char(line_above[char_idx])
            or is_special_char(line_above[char_idx + 1])
            or is_special_char(line_below[char_idx])
            or is_special_char(line_below[char_idx + 1])):
                return True

        elif char_idx == (len(current_line) - 1): # eind
            if (is_special_char(current_line[char_idx - 1])
            or is_special_char(line_above[char_idx])
            or is_special_char(line_above[char_idx - 1])
            or is_special_char(line_below[char_idx] )
            or is_special_char(line_below[char_idx - 1] )):
                return True

        elif char_idx > 0 and char_idx < (len(current_line) - 1):
            if (is_special_char(current_line[char_idx + 1])
            or is_special_char(current_line[char_idx - 1])
            or is_special_char(line_above[char_idx])
            or is_special_char(line_above[char_idx - 1] )
            or is_special_char(line_above[char_idx + 1] )
            or is_special_char(line_below[char_idx])
            or is_special_char(line_below[char_idx - 1] )
            or is_special_char(line_below[char_idx + 1])):
                return True

        return False

    else:
        print("Invalid linenr {}".format(current_line_nr))
        return False

def find_part_numbers_w_idx_v2(filename):
    with open(filename, 'r') as f:
        lines = f.readlines()

    line_count = len(lines)

    result = []
    current_number_str = ""

    result_sum = 0

    for idx, line in enumerate(lines):
        result_line = []
        save_curr_num = False
        for idy, c in enumerate(line):
            if c.isdigit():
                current_number_str += c
                #print(f"Checking digit: {lines[idx][idy]} line: {idx} index {idy}")
                if has_sticky_symbol(lines, idx, idy):
                    #print(f"Has sticky! {lines[idx][idy]} line {idx} index {idy}")
                    save_curr_num = True
            elif c.isdigit() and idy == (len(line) - 1):
                if current_number_str != "":
                    if save_curr_num:
                        current_number = int(current_number_str)

                        print(f"{current_number}, ")
                        start_num_idx = idy - len(current_number_str)

                        result.append((current_number, start_num_idx))
                        result_line.append((current_number, start_num_idx))
                        result_sum += current_number

                        save_curr_num = False

                    current_number_str = ""
            else:
                if current_number_str != "":
                    if save_curr_num:
                        current_number = int(current_number_str)

                        print(f"{current_number}, ")
                        start_num_idx = idy - len(current_number_str)

                        result.append((current_number, start_num_idx))
                        result_line.append((current_number, start_num_idx))
                        result_sum += current_number

                        save_curr_num = False

                    current_number_str = ""

        print("Line {} {}\n".format(idx, result_line))

    print("v2 sum {}".format(result_sum))
    return result

def day_three_part_one_v2(filename):
    output = find_part_numbers_w_idx_v2(filename)
    answer = sum([value for value, _ in output])
    return answer

ans = day_three_part_one_v2(filename)
print("ans v2: ", ans)

File: 2023/src/three/day_three.rs
use std::{fs::File, io::{BufReader, BufRead}, path::Path};

use crate::filereader;


fn is_adjacent(line: &str, start_idx: i32, number_length: i32) -> bool
{
    let min_bound = start_idx - 1;
    let max_bound = start_idx + number_length;
    for (idx, c) in line.char_indices()
    {
        if c != '.'  && !c.is_digit(10) 
        {
            // ...|*....|...
            // ...|.123(|...
            // ...|....&|...
            // Alles tussen de lijnen geldig
            if idx as i32 >= min_bound && idx as i32 <= max_bound // misschien + 1
            {
                //println!("Special char! {}", c);
                return true;
            }
        }
    }
    return false;
}

fn is_symbol_around(lines: &mut Vec<String>, line_cnt: i32, line_nr: i32, start_idx: i32, number_str: &str) -> bool
{
    let number_length = number_str.len();

    //println!("BGD LINECOUNT {}", line_cnt);
    //println!("BGD lnnr {}", line_nr);

    // regel 4: 777 niet gepakt

    if line_nr == 0
    {
        //print!("{}, ", line_nr);
        // check only below
        let current_line = lines.get(line_nr as usize).clone().unwrap();
        let line_below = lines.get(line_nr as usize + 1).clone().unwrap();
        return is_adjacent(&current_line, start_idx, number_length as i32) || is_adjacent(&line_below, start_idx, number_length as i32)
    }
    else if line_nr == line_cnt - 1
    {
        //print!("{}, ", line_nr);
        // check only above
        let line_above = lines.get(line_nr as usize - 1).unwrap();
        let current_line = lines.get(line_nr as usize).clone().unwrap();
        return is_adjacent(&line_above, start_idx, number_length as i32) || is_adjacent(&current_line, start_idx, number_length as i32);
    }
    else if line_nr > 0 && line_nr < line_cnt
    {
        //print!("{}, ", line_nr);
        // check under and above
        let line_above = lines.get(line_nr as usize - 1).unwrap();
        let current_line = lines.get(line_nr as usize).unwrap();
        let line_below = lines.get(line_nr as usize + 1).unwrap();

        return is_adjacent(line_below, start_idx, number_length as i32)
        || is_adjacent(current_line, start_idx, number_length as i32)
        || is_adjacent(line_above, start_idx, number_length as i32);
    }
    else 
    {
        println!("Invalid linenr {}", line_nr);
        return false;
    }
}

fn find_part_numbers_w_idx(filename: &str) -> Vec<(i32, usize)>
{
    let mut lines_orig = filereader::read_lines(filename).unwrap();
    //let mut lines_ref = filereader::read_lines(filename).unwrap();
    //let mut lines_ref2 = filereader::read_lines(filename).unwrap();

    let file = File::open(Path::new(filename)).unwrap();
    let file2 = File::open(Path::new(filename)).unwrap();
    let reader = BufReader::new(&file);
    let reader2 = BufReader::new(&file2);
    let lines: Vec<_> = reader.lines().collect::<Result<_, _>>().unwrap();
    let mut lines2: Vec<_> = reader2.lines().collect::<Result<_, _>>().unwrap();

    //println!("LINES {:?}", lines2);

    let line_count = lines2.len() as i32;

    //println!("BGD LINECOUNT {}", line_count);

    let mut result = Vec::new();
    let mut current_number_str = String::new();

    let mut current_line_nr: i32 = 0;

    let mut last_idx = 0;

    let mut line_cntr = 0;

    let mut result_sum = 0;
    for line in lines
    {
        line_cntr += 1;
        let mut result_line = Vec::new();
        for (idx, c) in line.char_indices()
        {
            if let Some(digit) = c.to_digit(10) 
            {
                // sowieso pushen
                current_number_str.push(c);
            } 
            else
            {
                if !current_number_str.is_empty() 
                {
                    //result.push((current_number_str.parse::<i32>().unwrap(), idx));
                    let current_number = current_number_str.parse::<i32>().unwrap();

                    // check around
                    
    	            let start_num_idx = idx as i32 - current_number_str.len() as i32;
                    //println!("n start idx: {}", start_num_idx);

                    // check per char eromheen
                    if is_symbol_around(&mut lines2, line_count, current_line_nr, start_num_idx, current_number_str.as_str())
                    {
                        result.push((current_number, start_num_idx as usize));
                        result_line.push((current_number, start_num_idx as usize));
                        result_sum += current_number;
                    }
                    current_number_str.clear();
                }
            }
            last_idx = idx;
        }
        
        println!("Line {} {:?}\n", line_cntr, result_line);
        // if !current_number_str.is_empty() 
        // {
        //     result.push((current_number_str.parse::<i32>().unwrap(), last_idx));
        // }
        current_line_nr += 1;
    }

    // 558036 te laag
    // 560570 te laag

    //println!("{:?}", result);
    println!("{:?}", result_sum);
    return result;

    
}

pub fn day_three_part_one(filename: &str) -> i32
{
    let output: Vec<(i32, usize)> = find_part_numbers_w_idx(filename);
    let answer = output.iter().map(|&(value, _)| value).sum();
    return answer;
}